File: src/main/java/uk/yermak/audiobookconverter/ArtWorkImage.java
package uk.yermak.audiobookconverter;

import javafx.embed.swing.SwingFXUtils;
import javafx.scene.image.Image;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.lang.invoke.MethodHandles;

public class ArtWorkImage implements ArtWork {
    private final Image image;
    private final Logger logger;
    private ArtWork bean;

    public Image image() {
        return this.image;
    }

    private Logger logger() {
        return this.logger;
    }

    private ArtWork getBean() {
        if (this.bean != null) {
            return this.bean;
        }

        BufferedImage bImage = SwingFXUtils.fromFXImage(this.image(), null);
        String poster = Utils.getTmp(this.image().hashCode(), this.image().hashCode(), "png");

        try {
            File posterFile = new File(poster);
            posterFile.deleteOnExit();
            ImageIO.write(bImage, "png", posterFile);
            long crc32 = Utils.checksumCRC32(posterFile);
            this.bean = new ArtWorkBean(poster, crc32);
        } catch (IOException var7) {
            this.logger().error("Error during parallel conversion", var7);
            var7.printStackTrace();
        }

        return this.bean;
    }

    public long getCrc32() {
        return this.getBean().getCrc32();
    }

    public String getFileName() {
        return this.getBean().getFileName();
    }

    @Override
    public boolean matchCrc32(long crc32) {
        return getCrc32() == crc32;
    }

    public ArtWorkImage(final Image image) {
        this.image = image;
        this.logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
        this.bean = null;
    }
}

        

File: src/main/scala/uk/yermak/audiobookconverter/ArtWorkImage.scala
package uk.yermak.audiobookconverter

import java.io.{File, IOException}
import java.lang.invoke.MethodHandles

import javafx.embed.swing.SwingFXUtils
import javafx.scene.image.Image
import javax.imageio.ImageIO
import org.slf4j.{Logger, LoggerFactory}

/**
  * Created by yermak on 16-Nov-18.
  */
class ArtWorkImage(var image: Image) extends ArtWork {
  private val logger = LoggerFactory.getLogger(MethodHandles.lookup.lookupClass)

  private var bean: ArtWork = null

  private def getBean: ArtWork = {
    if (bean != null) bean
    val bImage = SwingFXUtils.fromFXImage(image, null)
    val poster = Utils.getTmp(image.hashCode, image.hashCode, "." + "png")
    try {
      val posterFile = new File(poster)
      posterFile.deleteOnExit()
      ImageIO.write(bImage, "png", posterFile)
      val crc32 = Utils.checksumCRC32(posterFile)
      bean = new ArtWorkBean(poster, "png", crc32)
    } catch {
      case e: IOException =>
        logger.error("Error during parallel conversion", e)
        e.printStackTrace()
    }
    bean
  }

  override def getCrc32: Long = getBean.getCrc32

  def setCrc32(crc32: Long): Unit = getBean.setCrc32(crc32)

  override def getFileName: String = getBean.getFileName

  override def setFileName(fileName: String): Unit =
    getBean.setFileName(fileName)

  override def getFormat: String = getBean.getFormat

  override def setFormat(format: String): Unit = getBean.setFormat(format)
}
